hax/hax/hax.py

```diff
@@ -100,18 +100,18 @@ def main():
     consumer = _run_qconsumer_thread(q, motr, herald, util)
 
     try:
-        motr.start(cfg.hax_ep,
-                   process=cfg.hax_fid,
-                   ha_service=cfg.ha_fid,
-                   rm_service=cfg.rm_fid)
+        motr_runnable = motr.start(cfg.hax_ep,
```

Comment:
Up to this moment we had only one thread that "owns" everything related to Motr - ConsumerThread. If ConsumerThread is running, hax can communicate to Motr.

#1438 states that this picture is not balanced though: rconfc stuff happens outside of ConsumerThread but implicitly depends on ConsumerThread (because that guy owns Motr context which is a must for rconfc magic to happen).

So per my understanding it would be more logical to keep all Motr-related stuff in a single place (i.e. rconfc and motr context together). But you added one more thread in addition to ConsumerThread and StatsUpdater 🤔 

Do you think your approach with MotrRunnable is better?

Reply:
Because rconfc_start is a blocking operation (requires motr processes to be started, atleast confd) we need a non-blocking context, hence the thread. Effectively the thread doesn't do anything but just starts rconfc and sleeps until hax process terminates.
You do have a point that this active thread can give an impression of multiple motr active processing contexts, I did think of adding a separate message type to start rconfc once all the motr processes have started but rconfc start can still block.
May be the name motrrunnable is a bit misnomer and can be renamed to may be just rconfcManager or something.

Reply:
@mssawant 
> Because rconfc_start is a blocking operation (requires motr processes to be started, atleast confd) we need a non-blocking context, hence the thread. 

I don't think it is a problem. I think it is OK if a thread blocks until rconfc_start starts. In other words, if a thread needs to work with Motr (or with Spiel API), it needs to have the corresponding context initialized. So it must wait until it is initialized just because it can't do anything meaningful without the context anyway. 

So it seems like this splitting to a number of threads will make the things more complicated (inter-thread communication, like "a separate message type to start rconfc once all the motr processes have started" in your example) without a real benefit.

Reply:
I'm just trying to avoid adding extra [accidental complexity](https://medium.com/background-thread/accidental-and-essential-complexity-programming-word-of-the-day-b4db4d2600d4).

Reply:
@knekrasov, If not in a separate thread, do you suggest to start rconfc in consumer thread and block consumer thread then until rconfc starts? I see a deadlock there as hax will not be able to reply to entrypoint and nvec requests which will block motr confd and ioservices.

Reply:
@mssawant 
Hm, good point about the deadlock.

What if we use a compromise solution like this:
![image](https://user-images.githubusercontent.com/69299710/104458991-2aca1a00-55bd-11eb-9734-a867e0f254f6.png)

In short, 
1. rconfc is initialized in a separate thread to `qconsumer`, that thread is called `rconfc_stater` at the diagram
2. `rconfc_starter` is short-lived, it exits as soon as rconfc is started. Its aim is effectively to start rconfc asynchronously to `qconsumer`
3. overall Motr context is still owned by one thread - `qconsumer`.

Reply:
Agree, have updated the patch with necessary changes. Thank you.